### ISSUE TITLE
Nginx  - Some problems

### ISSUE THREAD

**Issue by Draleg** [shaarli/Shaarli#1159]
Hello all,

I want my old server to a Raspberry Pi 3B+.
In that case, I have installed NGINX and PHP-FPM.

First of all "**link**" is my folder name for shaarli.

When I test I have some problems:

### 1) I do a copy/past from my old server to my RPI.
Shaarli is reachable but impossible to use my login/password.
So I remove "data/config.json.php" and the second problem appears.

However I don't know why my login/password is rejected because it's the good one...

### 2) Redirections fails!
When a remove "data/config.json.php" I have a bad redirection like this:
"_/link/?test_session"
I supposed that when a remove the config.json.php a other redirection is try but I can't see it and I can't reproduce it with my local IP.
Because when a go to ?test_session with all the good URL it's a warning : 

"Sessions do not seem to work correctly on your server.
Make sure the variable "session.save_path" is set correctly in your PHP config, and that you have write access to it.
It currently points to /var/lib/php/sessions.
On some browsers, accessing your server via a hostname like 'localhost' or any custom hostname without a dot causes cookie storage to fail. We recommend accessing your server via it's IP address or Fully Qualified Domain Name.
"
I read the documentation:
https://shaarli.readthedocs.io/en/master/Server-configuration/#nginx

### And a I see :

      location /link/ {
            try_files $uri /link/index.php$is_args$args;
            access_log  /var/log/nginx/shaarli.access.log;
            error_log   /var/log/nginx/shaarli.error.log;
        }

      location ~ (index)\.php$ {
            try_files $uri =404;
            fastcgi_split_path_info ^(.+\.php)(/.+)$;
            fastcgi_pass   unix:/var/run/php-fpm/php-fpm.sock;
            fastcgi_index  index.php;
            include        fastcgi.conf;
        }

### And also:
      location /link/ {
            # Slim - rewrite URLs
            try_files $uri /index.php$is_args$args;

            access_log  /var/log/nginx/shaarli.access.log;
            error_log   /var/log/nginx/shaarli.error.log;
        }

I try all the modification and one by one but redirection is always fail.

I suppose that all my problem is about this fail redirection, and I don't know Nginx like Apache... But Nginx is really recommanded for RPi...

I don't know if it's important, but the redirection problem is also with the button "RSS Feed" and "ATOM Feed" and not with the other button (I suppose that the other one are "hard linked", I don't verify that).
This is the link for "RSS Feed" = **http://_/link/?do=rss**
This is the link for "ATOM Feed" = **http://_/link/?do=atom**

I try also this configuration without SSL configuration just 80 and with all the good replace shaarli by link:
https://framacloud.org/fr/auto-hebergement/services_web.html#partage-de-liens-avec-shaarli-%E2%98%85
But it's also fail...

So somebody to help me ?

Thank's in advance.

**Comment by Draleg** [comment]
Hello all,

So the login problem est solved... I don't know how but its ok now (I do several RM,CP,CHOWN so it's very difficult to know why before it was fail and now it's ok)...

The redirect problem is localised ! It's the serveur_name variable in default file from NGINX.
But's its a problem ... I use a dynamic DNS with a redirecting port AND I use my shaarli also in my local network.
And when I follow instruction and I put "localhost" as server_name my redirecting link is : 
**localhost/link/?do=rss** 
There is no interpretation, NGINX put the string name from server_name and that's all!

I try some configuration with several name:
server_name ddns localip;
But also fails... NGINX choose firstone and that's all...

I found some documentation with reversing proxy .. all fail !

If there is a NGINX expert lol... I'm need it...

Thank's all.

**Comment by ArthurHoaro** [comment]
Hi,
If I understand correctly, you want to access Shaarli from 2 distinct domain. In that case, you will need 2 distinct virtual hosts in nginx, because Shaarli uses the `server_name` to determine its root URL and, for example, attach its cookies to it.

You can also use the dyndns domain everywhere, and optionally add it to your `/etc/hosts` file.

**Comment by virtualtam** [comment]
> I don't know Nginx like Apache... But Nginx is really recommended for RPi

There is nothing wrong with running Apache on a rPI, especially given how beefy recent models have become. Sure, it might eat a bit more resources than Nginx, but if you're comfortable with it, go for it!

We provide a working Nginx configuration with the Docker images, as well as a recommended setup in the docs. It won't fit each and every case though, and we can't support each and every server combination either.

Now if you want to try Nginx, you need to consider the following points:
- the Nginx processes are owned by a given user:group, usually `http:http` (this might differ according to the distribution you used)
- the PHP-FPM processes are also owned owned by a given user:group
- Nginx directly serves static resources (assets, CSS)
- Nginx forwards PHP requests to PHP-FPM

This means:
- the Shaarli files must be owned by the PHP-FPM user:group, which must also have access to:
    - read configuration, logs, PHP files, etc.
    - write data to the data and cache directories
- Nginx must be able to:
    - read static resources
    - communicate with PHP-FPM

For example:
- Shaarli files are owned by `http:http`
- file permissions are set accordingly (see docs for detailed file/folder permissions)
- PHP-FPM is running as `http:http`
- Nginx is running as `http:http`

This should provide you with a working instance served on localhost.

As for external access, it is not clear from your posts:
- how the rPI is configured
- which distribution is used
- which packages are installed
- how the network is configured
- how Nginx is configured
- how PHP-FPM is configured
- whether Shaarli is served and/or accessed behind a reverse HTTP proxy

With this in mind, I would suggest updating and formatting your post(s) so these points are clearly identified; this will be helpful to:
1. check your configuration,
2. understand what's going on,
3. maybe root-cause the issue,
4. future reference should other users face the same situation.

**Comment by Draleg** [comment]
Hello, all :-) .
And thank's for your comments.

virtualtam:
There is nothing wrong with running Apache on a rPI, especially given  how beefy recent models have become. Sure, it might eat a bit more  resources than Nginx, but if you're comfortable with it, go for it!
--

That's what I did last night lol... I remove nginx/nginx-common/php-fpm and I install apache2/php.
And everything is alright! 
I'm the only user from my webserver (just because it's really for my that I built it)...
So after thinking about that, I really sure that there is no big deal to use Apache instead of Nginx in my case.


Thank's .
